### PR TITLE
Add SM server & VQSR

### PR DESCRIPTION
**Pull request by vladsavelyev**
Update the loading pipeline to support VQSR, and interact with the sample-metadata server.

1. `batch_seqr_loader/batch_workflow.py` is run with parameters `--sm-db-name` (defaults to `seqr`) and `--seqr-dataset` (e.g. acute-care).

2. Pipeline pulls data for all available samples, and finds the latest complete `analysis` entries by type.

3. For each sample:
  a) if a `cram` completed analysis does not exist, the pipeline checks the `reads` value in the sample `metadata`, and if it is can be parsed as a CRAM file or a set of FASTQ pairs (one per lane), it records a `cram` `queued` analysis, submits a Batch job with the bazam -> BWA -> marksomadup pipeline. It also submits a job that updates the status to `in-progress`, which runs before BWA; and a job that updates it to `completed` after BWA.
  b) if a `gvcf` completed analysis does not exist, it records a `gvcf` `queued`analysis and submits SplitIntervals, followed by a HaplotypeCaller job for each interval, then GatherGVCF to merge the intervals, and finally adds postprocessing jobs: ReblockGVCFs and subset to no-alt contigs. It also submits a job that updates the status to `in-progress`, which runs before the SplitIntervals job; and a job that updates it to `completed` dependent on the last step of GVCF postprocessing.
 
4. If a completed `joint-calling` analysis does not exist, it records a `queued``joint-calling` analysis, and submits a set of jobs: SplitIntervals, update or create GenomicsDB per interval, GnarlyGenotyper and MakeSitesOnlyVcf per interval, then gathers VCFs, and submits VQSR jobs: IndelsVariantRecalibrator, SNPsVariantRecalibrator (per interval), ApplyRecalibration, gather VCFs, VariantEval. The first job depends on a job that update the analysis status to `in-progress`, and the last is followed by a job that updates it to `completed`.

5. A dataproc job is called to export the VCF to a matrix table and annotate it. 

6. Another dataproc job is called to create a ES index.

Test:
https://batch.hail.populationgenomics.org.au/batches/3996
https://batch.hail.populationgenomics.org.au/batches/4026
https://batch.hail.populationgenomics.org.au/batches/4133

TODO:
- [ ] Annotate separately for each seqr-dataset
- [x] Apply VQSR filters
- [ ] Read pedigree from the SM server

### CODE REVIEW

File: batch_seqr_loader/scripts/batch_index_bwa.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
```
Comment (lgruen) [inline batch_seqr_loader/scripts/batch_index_bwa.py]: Was this meant to go in the `/test` subdirectory?

File: batch_seqr_loader/test/test_sm.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+def _jc_pipeline_submit_analyses():
+    """
+    Add or update analyses. Iterate over completed analyses,
+    and submit next-step analyses
+    """
+
+    # If there are incomplete analyses, throw an error
+    aapi = AnalysisApi()
+    print('aapi = AnalysisApi()')
+    # analyses = aapi.get_incomplete_analyses(project=PROJ)
```
Comment (lgruen) [inline batch_seqr_loader/test/test_sm.py]: Should all of this be commented out?

File: batch_seqr_loader/test/test_sm.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+def test_simulate_joint_calling_pipeline():
+    """
+    Simulates events of the joint-calling workflow
+    """
+
+    test_run_id = '1B9R9K'
+    sample_ids = 'CPG11775, CPG11783, CPG11791'.split(', ')
+
+    # Unique test run ID to avoid clashing with previous test run samples
```
Comment (lgruen) [inline batch_seqr_loader/test/test_sm.py]: Ditto.

File: batch_seqr_loader/utils.py [hunk 4]
```diff
@@ -1,9 +1,54 @@
...
+logger = logging.getLogger(__file__)
+logging.basicConfig(format='%(levelname)s (%(name)s %(lineno)s): %(message)s')
+logger.setLevel(logging.INFO)
+
+
+GATK_VERSION = '4.2.0.0'
+GATK_IMAGE = (
+    f'australia-southeast1-docker.pkg.dev/cpg-common/images/gatk:{GATK_VERSION}'
+)
+PICARD_IMAGE = f'us.gcr.io/broad-gotc-prod/picard-cloud:2.23.8'
```
Comment (lgruen) [inline batch_seqr_loader/utils.py]: Should we copy these US images, to avoid egress costs?

File: batch_seqr_loader/utils.py [hunk 5]
```diff
@@ -27,3 +72,116 @@ def file_exists(path: str) -> bool:
...
+        logger.info(f'File {fpath} exists and will be overwritten')
+        return False
+    else:
+        logger.info(f'Reusing existing {fpath}. Use --overwrite to overwrite')
+        return True
+
+
+def hash_sample_names(sample_names: Iterable[str]) -> str:
+    """
+    Return a unique hash string from a from a set of strings
```
Comment (lgruen) [inline batch_seqr_loader/utils.py]:
```suggestion
    Return a unique hash string from a set of strings
```

File: batch_seqr_loader/utils.py [hunk 6]
```diff
@@ -27,3 +72,116 @@ def file_exists(path: str) -> bool:
...
+    j = b.new_job(f'SM: update {analyais_type} to {status}')
+    j.image(SM_IMAGE)
+    j.command(
+        f"""
+set -o pipefail
+set -ex
+
+export GOOGLE_APPLICATION_CREDENTIALS=/gsa-key/key.json
+gcloud -q auth activate-service-account --key-file=$GOOGLE_APPLICATION_CREDENTIALS
+export SM_DEV_DB_PROJECT=vladdev
```
Comment (lgruen) [inline batch_seqr_loader/utils.py]: Still pointing to `vladdev` here.

File: batch_seqr_loader/batch_workflow.py [hunk 7]
```diff
@@ -195,142 +131,167 @@ def main(
...
+    web_bucket_suffix,  # pylint: disable=unused-argument
+    seqr_dataset_name,
+    seqr_dataset_version,  # pylint: disable=unused-argument
+    overwrite: bool,
+    prod: bool,
+    vep_block_size,
+):
+    genomicsdb_bucket = f'{out_bucket}/genomicsdbs'
+    # fingerprints_bucket = f'gs://cpg-seqr-{output_suffix}/datasets/{seqr_dataset_name}/{seqr_dataset_version}/fingerprints'
+    # web_bucket = f'gs://cpg-seqr-{web_bucket_suffix}/datasets/{seqr_dataset_name}/{seqr_dataset_version}'
```
Comment (lgruen) [inline batch_seqr_loader/batch_workflow.py]:
```suggestion
```

File: batch_seqr_loader/batch_workflow.py [hunk 8]
```diff
@@ -195,142 +131,167 @@ def main(
...
+
+    samples = sapi.get_all_samples(project=sm_db_name)
+    samples = [s for s in samples if not s.external_id.startswith('NA12878-from')]
```
Comment (lgruen) [inline batch_seqr_loader/batch_workflow.py]: What's this about?

File: batch_seqr_loader/batch_workflow.py [hunk 9]
```diff
@@ -195,142 +131,167 @@ def main(
...
             f'--dest-mt-path {annotated_mt_path} '
-            f'--bucket {join(work_bucket, "seqr_load")} '
-            + (f'--disable-validation ' if disable_validation else '')
-            + (f'--make-checkpoints ' if make_checkpoints else '')
-            + (f'--remap-tsv ' if remap_path else '')
-            + (f'--subset-tsv ' if subset_path else '')
-            + (f'--vep-block-size ' if vep_block_size else ''),
+            f'--bucket {join(out_bucket, "seqr_load")} '
+            '--disable-validation '
+            '--make-checkpoints ' + (f'--vep-block-size ' if vep_block_size else ''),
```
Comment (lgruen) [inline batch_seqr_loader/batch_workflow.py]: Will 8h be enough?

File: batch_seqr_loader/batch_workflow.py [hunk 10]
```diff
@@ -195,142 +131,167 @@ def main(
...
         b,
         f'batch_seqr_loader/scripts/load_to_es.py '
         f'--mt-path {annotated_mt_path} '
-        f'--es-index {dataset_name}-{dataset_version} '
+        f'--es-index {seqr_dataset_name}-{seqr_dataset_name} '
         f'--es-index-min-num-shards 1 '
         f'--genome-version GRCh38 '
-        f'{"--prod" if output_namespace == "main" else ""}',
+        f'{"--prod" if prod else ""}',
         max_age='8h',
```
Comment (lgruen) [inline batch_seqr_loader/batch_workflow.py]: Ditto.

File: batch_seqr_loader/utils.py [hunk 11]
```diff
@@ -1,9 +1,54 @@
...
+import hailtop.batch as hb
+from hailtop.batch import Batch
+from hailtop.batch.job import Job
+
+logger = logging.getLogger(__file__)
+logging.basicConfig(format='%(levelname)s (%(name)s %(lineno)s): %(message)s')
+logger.setLevel(logging.INFO)
+
+
+GATK_VERSION = '4.2.0.0'
```
Comment (lgruen) [inline batch_seqr_loader/utils.py:20]: Should we use 4.2.1.0, to take advantage of the better GCS integration for GenomicsDB?

> GenomicsDB update: GATK has now moved to the new version of GenomicsDB (version 1.4.1), with native cloud support (previously, it relied on the slower GCS Connector library). At this point, it should basically be indistinguishable from using a network file system.

**Review by lgruen — CHANGES_REQUESTED** [review]

File: batch_seqr_loader/batch_workflow.py (hunk 8 above)
Comment (vladsavelyev) [reply batch_seqr_loader/batch_workflow.py]: That's for testing. Removed!

File: batch_seqr_loader/batch_workflow.py (hunk 9 above)
Comment (vladsavelyev) [reply batch_seqr_loader/batch_workflow.py]: For larger datasets, probably not...

File: batch_seqr_loader/utils.py (hunk 11 above)
Comment (vladsavelyev) [reply batch_seqr_loader/utils.py:20]: I haven't tested that yet! For the following PR :)

**Review by lgruen — APPROVED** [review]

File: batch_seqr_loader/test/sm_test_add_data.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+
+sapi = SampleApi()
+s1 = NewSample(
+    external_id='210303_A00692_0190_ML211637',
+    type='blood',
+    meta={
+        'reads': [
+            [
+                'gs://cpg-acute-care-test-upload/cpg_acute_20210729_084835/210303_A00692_0190_ML211637_21W000380-FAM001171_MAN-20210303_NEXTERAFLEXWGS_L001_R1.fastq.gz',
```
Comment (illusional) [inline batch_seqr_loader/test/sm_test_add_data.py]: This isn't the format that reads are in though right?

File: batch_seqr_loader/test/test_sm.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+
+    aapi = AnalysisApi()
+    print('aapi = AnalysisApi()')
+
+    # Get the list of latest complete analyses
+    latest_complete_analyses = aapi.get_latest_complete_analyses(project=PROJ)
+    print(f'Latest complete analyses: {latest_complete_analyses}')
+    latest_by_type_and_sids = defaultdict(list)
+    for a in latest_complete_analyses:
+        a_s_ids = sample_id_format(a['sample_ids'])
```
Comment (illusional) [inline batch_seqr_loader/test/test_sm.py]: The SM server should be doing this operation, I've sent a you a patch on slack :)

File: batch_seqr_loader/batch_workflow.py [hunk 14]
```diff
@@ -9,102 +9,63 @@
...
-    '--data-to-realign',
-    'data_to_realign',
-    multiple=True,
-    help='Glob path with input BAM/CRAM/FASTQ files that need re-alignment. '
-    'For BAMs and CRAMs, corresponding bai/crai indices are expected to exist. '
-    'Files will be (re-)aligned with BWA, and variants will be called with '
-    'GATK HaplotypeCaller',
+    '--sm-db-name',
+    'sm_db_name',
+    default='seqr',
```
Comment (illusional) [inline batch_seqr_loader/batch_workflow.py]: We could remove this default given we're creating SM projects for each subproject.

File: batch_seqr_loader/batch_workflow.py [hunk 15]
```diff
@@ -1146,29 +1612,48 @@ def _add_final_gather_vcf_step(
...
-    :return: a string hash
-    """
-    for sn in sample_names:
-        assert ' ' not in sn, sn
-    return hashlib.sha224(' '.join(sorted(sample_names)).encode()).hexdigest()
+    error_msg = (
+        f'     ERROR: the "reads" meta data must be a list of '
+        f'str/list (for FASTQ), or a BAM or CRAM path. Got: {reads_data}'
+    )
+    if isinstance(reads_data, str):
```
Comment (illusional) [inline batch_seqr_loader/batch_workflow.py]: This isn't the format for reads in the sample metadata. They're formed CWL File objects, eg:

```json
"reads": [
    {
        "location": "gs://<bucket>/<sampleId>.bam",
        "basename": "<sampleId>.bam",
        "class": "File",
        "checksum": "md5:<md5-sum>",
        "size": 0, # in bytes
        "secondaryFiles": [
            {
                "location": "gs://<bucket>/<sampleId>.bam.bai",
                "basename": "<sampleId>.bam.bai",
                "class": "File",
                "checksum": null,
                "size": 0
            }
        ]
    }
],
"reads_type": "bam"
```

OR 

```json
"reads": [
    [
        {
            "location": "gs://<bucket>/{sampleId}_L001_R1.fastq.gz",
            "basename": "{sampleId}_L001_R1.fastq.gz",
            "class": "File",
            "checksum": "md5:checksum",
            "size": 13939740017
        },
        {
            "location": "gs://<bucket>/{sampleId}_L001_R2.fastq.gz",
            "basename": "{sampleId}_L001_R2.fastq.gz",
            "class": "File",
            "checksum": "md5:checksum",
            "size": 14710616940
        }
    ],
    [
        {
            "location": "gs://<bucket>/{sampleId}_L002_R1.fastq.gz",
            "basename": "{sampleId}_L002_R1.fastq.gz",
            "class": "File",
            "checksum": "md5:checksum",
            "size": 14061476086
        },
        {
            "location": "gs://<bucket>/{sampleId}_L002_R2.fastq.gz",
            "basename": "{sampleId}_L002_R2.fastq.gz",
            "class": "File",
            "checksum": "md5:checksum",
            "size": 14824459848
        }
    ]
],
"reads_type": "fastq"
```

**Review by illusional — CHANGES_REQUESTED** [review]
I don't see pedigree checks in there as a TODO, I might've missed it though, happy to help integrate that with SM server :)

I've requested changes for
- sample.meta format
- analysis sample IDs, happy to chat about this as well.

File: batch_seqr_loader/batch_workflow.py (hunk 15 above)
Comment (vladsavelyev) [reply batch_seqr_loader/batch_workflow.py]: Yeah, it just worked for my fake test data, forgot to update it to the real entries! Did now :)

File: batch_seqr_loader/test/test_sm.py (hunk 13 above)
Comment (vladsavelyev) [reply batch_seqr_loader/test/test_sm.py]: Will leave it for now, and will update in a following PR :)

File: batch_seqr_loader/test/sm_test_add_data.py (hunk 12 above)
Comment (vladsavelyev) [reply batch_seqr_loader/test/sm_test_add_data.py]: Yeah, will update!

**Comment by vladsavelyev** [comment]
@illusional, addressed the `sample.meta.reads_data` format question, but will have to wait for the sample-metadata update to address the analysis sample IDs.

File: batch_seqr_loader/test/sm_test_add_data.py (hunk 12 above)
Comment (vladsavelyev) [reply batch_seqr_loader/test/sm_test_add_data.py]: Done :)

File: batch_seqr_loader/batch_workflow.py (hunk 14 above)
Comment (vladsavelyev) [reply batch_seqr_loader/batch_workflow.py]: Leaving this option as `--test-sm-db-name` just to support testing with `vladdev`. The default value is `acute-care` for now, but will rename after the DB reorganisation :)

File: batch_seqr_loader/test/test_sm.py (hunk 13 above)
Comment (vladsavelyev) [reply batch_seqr_loader/test/test_sm.py]: Updated!

**Review by illusional — APPROVED** [review]

**Pull request merged**